src/Stateless/StateMachine.Async.cs

```diff
@@ -162,12 +162,74 @@ async Task InternalFireAsync(TTrigger trigger, params object[] args)
                 case FiringMode.Queued:
                     await InternalFireQueuedAsync(trigger, args);
                     break;
+                case FiringMode.Serial:
+                    await InternalFireSerialAsync(trigger, args);
+                    break;
                 default:
                     // If something is completely messed up we let the user know ;-)
                     throw new InvalidOperationException("The firing mode has not been configured!");
             }
         }
 
+        /// <summary>
+        /// Queue events and then fire in order.
+        /// If only one event is queued, this behaves identically to the non-queued version.
+        /// </summary>
+        /// <param name="trigger">  The trigger. </param>
+        /// <param name="args">     A variable-length parameters list containing arguments. </param>
+        async Task InternalFireSerialAsync(TTrigger trigger, params object[] args) {
+
+            lock (_serialModeLock) 
```

Comment:
Using a regular lock in an async method isn't a best practice as it blocks the thread during the wait for the lock. Consider using `SemaphoreSlim(1,1)` that supports both sync and async waiting instead.